### PR TITLE
Target optional for seq2seq

### PR DESCRIPTION
**Pull request by Kelina**
Gold target sequence not needed for seq2seq generation, and only used *during training* if available.

### CODE REVIEW

File: jiant/modules/seq2seq_decoder.py [hunk 1]
```diff
@@ -44,8 +44,8 @@ def __init__(
 
         # We need the start symbol to provide as the input at the first timestep of decoding, and
         # end symbol as a way to indicate the end of the decoded sequence.
-        self._start_index = self.vocab.get_token_index(START_SYMBOL, self._target_namespace)
-        self._end_index = self.vocab.get_token_index(END_SYMBOL, self._target_namespace)
+        self._start_index = self.vocab.get_token_index("<SOS>", self._target_namespace)
```
Comment (W4ngatang) [inline jiant/modules/seq2seq_decoder.py]: is this different than `START_SYMBOL` ?

Comment (Kelina) [reply jiant/modules/seq2seq_decoder.py]: Yes.

Comment (W4ngatang) [reply jiant/modules/seq2seq_decoder.py]: Can you search the codebase that this won't break somewhere else ? I'm worried that in preprocessing or in task definitions, we defined the token differently.

File: jiant/modules/seq2seq_decoder.py [hunk 2]
```diff
@@ -185,10 +191,15 @@ def forward(
...
             target_mask = get_text_field_mask(target_tokens)
-            loss = self._get_loss(logits, targets, target_mask)
-            output_dict["loss"] = loss
             output_dict["target_mask"] = target_mask
+            relevant_logits = logits[:, : targets.shape[1] - 1, :].contiguous()
+            loss = self._get_loss(relevant_logits, targets, target_mask)
+            output_dict["loss"] = loss
+        else:
+            # This is needed in case no gold target sequence is available.
+            output_dict["loss"] = torch.tensor([-1.0])
```
Comment (W4ngatang) [inline jiant/modules/seq2seq_decoder.py]: fatal to use `None`?

File: jiant/modules/seq2seq_decoder.py [hunk 3]
```diff
@@ -185,10 +191,15 @@ def forward(
         output_dict = {"logits": logits}
 
         if target_tokens:
+            targets = target_tokens["words"]
             target_mask = get_text_field_mask(target_tokens)
-            loss = self._get_loss(logits, targets, target_mask)
-            output_dict["loss"] = loss
             output_dict["target_mask"] = target_mask
+            relevant_logits = logits[:, : targets.shape[1] - 1, :].contiguous()
```
Comment (W4ngatang) [inline jiant/modules/seq2seq_decoder.py]: what is going on here exactly? Shouldn't the targets be padded such that you don't need to use `targets.shape[1]` (the loss should ignore them if you pass in the pad value ) ?

File: jiant/tasks/seq2seq.py [hunk 4]
```diff
@@ -119,7 +119,8 @@ def get_metrics(self, reset=False):
         return {"perplexity": math.exp(avg_nll), "accuracy": acc}
 
     def update_metrics(self, logits, labels, tagmask=None):
-        self.scorer2(logits.max(2)[1], labels, tagmask)
+        relevant_logits = logits.max(2)[1][:, : labels.shape[1]]
```
Comment (W4ngatang) [inline jiant/tasks/seq2seq.py]: not sure what is going on with the shapes here, could you leave a comment ? (Also, could you use `logits.max(dim=2)` )

File: jiant/modules/seq2seq_decoder.py (hunk 2 above)
Comment (Kelina) [reply jiant/modules/seq2seq_decoder.py]: I think I tried that, yes. But it will also be changed in a follow-up PR.

File: jiant/modules/seq2seq_decoder.py (hunk 3 above)
Comment (Kelina) [reply jiant/modules/seq2seq_decoder.py]: Right; but in this PR the model generates up to maximum length (since the gold target sequence is not used). So we need to cut logits to the size of targets. The result for accuracy (which is the only thing we compute on this for now) is the same, since it is wrong anyways if <EOS> is not the last generated symbol.

File: jiant/tasks/seq2seq.py (hunk 4 above)
Comment (Kelina) [reply jiant/tasks/seq2seq.py]: Done.

File: jiant/modules/seq2seq_decoder.py (hunk 1 above)
Comment (Kelina) [reply jiant/modules/seq2seq_decoder.py]: Done.

File: jiant/modules/seq2seq_decoder.py (hunk 3 above)
Comment (W4ngatang) [reply jiant/modules/seq2seq_decoder.py]: sgtm
